File: src/main/java/core/procedures/strings/ListToString.java
package core.procedures.strings;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Type;

import java.util.List;

public final class ListToString extends AFn {

  public ListToString() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Type.SCMProperList.class}).build());
  }

  @Override
  public String getName() {
    return "list->string";
  }

  @Override
  public Object apply1(Object arg) {
    List cs = (List)arg;
    if (cs.isEmpty()) {
      return "";
    }
    StringBuilder sb = new StringBuilder(cs.size());
    for (Object c : cs) {
      if (!(c instanceof Character)) {
        throw new WrongTypeException(getName(), "Character", c);
      }
      sb.append(c);
    }
    return sb.toString();
  }
}

File: src/main/java/core/procedures/strings/ListToString.kt
package core.procedures.strings

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type

class ListToString : AFn<List<*>?, String>(name = "list->string", isPure = true, arity = Exactly(1),
                                           mandatoryArgsTypes = arrayOf(Type.ProperList::class.java)) {

    override operator fun invoke(arg: List<*>?) = when {
        arg!!.isEmpty() -> ""
        else -> arg.joinToString(separator = "") {
            (it as? Char)?.toString() ?: throw WrongTypeException(name, "Character", it)
        }
    }
}
